geomstats/geometry/base.py

```diff
@@ -366,12 +367,18 @@ def belongs(self, point, atol=gs.atol):
         belongs = self.embedding_space.belongs(point, atol)
         if not gs.any(belongs):
             return belongs
-        value = self.value
-        constraint = gs.isclose(self.submersion(point), value, atol=atol)
-        if value.ndim == 2:
-            constraint = gs.all(constraint, axis=(-2, -1))
-        elif value.ndim == 1:
-            constraint = gs.all(constraint, axis=-1)
+
+        submersed_point = self.submersion(point)
+        n = len(submersed_point.shape)
+        axis = (
+            tuple(-n + 1 + i for i in range(n - 1))
+            if gs.ndim(point) > len(self.shape)
+            else tuple(-n + i for i in range(n))
+        )
```

Comment:
Thanks @ymontmarin!

This solution fails for:

```python
space = Hyperboloid(3)

space.belongs(space.random_point())
```

Reply:
I think this may work in all cases:

```python
axis = tuple(range(n - len(self.shape), n)) if (n - len(self.shape)) >= 0 else ()
```

What do you think?

Reply:
Actually I'm not so sure this version increases readability (it may even make it harder to understand). Besides, we can easily achieve general batching by replacing 1 by `gs.ndim(point) - len(self.shape)` in the first range...

Reply:
a simplified and more readable version may be:

```python
n_batch = gs.ndim(point) - len(self.shape)
tuple(range(-n + n_batch, 0))
```

Reply:
After thinking about it, none of the two solutions are very universal.

`point` is of shape [*batch_shape, *self.shape]
`immersed_point` is of shape [*batch_shape, *sub_value_shape]

If the subermesion only give a real sub_value_shape = () otherwise we want the `all` on the sub_value_shape axis.

So maybe a more short and universal way that work for any batching, shape and output of submersion space would be:
```python
    def belongs(self, point, atol=gs.atol):
        """Evaluate if a point belongs to the manifold.

        Parameters
        ----------
        point : array-like, shape=[..., dim]
            Point to evaluate.
        atol : float
            Absolute tolerance.
            Optional, default: backend atol.

        Returns
        -------
        belongs : array-like, shape=[...,]
            Boolean evaluating if point belongs to the manifold.
        """
        belongs = self.embedding_space.belongs(point, atol)
        if not gs.any(belongs):
            return belongs

        submersed_point = self.submersion(point)
        constraint = gs.isclose(submersed_point, 0.0, atol=atol)

        n_sub_dim = len(submersed_point.shape) - len(point.shape) + len(self.shape)
        if n_sub_dim:
            constraint = gs.all(constraint, axis=tuple(range(-n_sub_dim, 0)))

        return gs.logical_and(belongs, constraint)
```

Reply:
My pessage was written before i saw your two messages ! I think it is almost the same :)

Reply:
I'm glad we came up with similar solutions independently. These ones I think are pretty universal. I'll merge now.